jax/_src/scipy/special.py

```diff
@@ -1089,3 +1089,379 @@ def sph_harm(m: jnp.ndarray,
+    return _eval_expint_k(A, B, x)
+
+
+def _expint5(x):
+    # 16 <= x <= 32
+    A = [
+        -2.458119367674020323359e-1,
+        -1.483382253322077687183e-1,
+        7.248291795735551591813e-2,
+        -1.348315687380940523823e-2,
+        1.342775069788636972294e-3,
+        -7.942465637159712264564e-5,
+        2.644179518984235952241e-6,
+        -4.239473659313765177195e-8,
+    ]
+    B = [
+        1.0,
+        -1.044225908443871106315e-1,
+        -2.676453128101402655055e-1,
+        9.695000254621984627876e-2,
+        -1.601745692712991078208e-2,
+        1.496414899205908021882e-3,
+        -8.462452563778485013756e-5,
+        2.728938403476726394024e-6,
+        -4.239462431819542051337e-8,
+    ]
+    return _eval_expint_k(A, B, x)
+
+
+def _expint6(x):
+    # 32 <= x <= 64
+    A = [
+        1.212561118105456670844e-1,
+        -5.823133179043894485122e-1,
+        2.348887314557016779211e-1,
+        -3.040034318113248237280e-2,
+        1.510082146865190661777e-3,
+        -2.523137095499571377122e-5,
+    ]
+    B = [
+        1.0,
+        -1.002252150365854016662e0,
+        2.928709694872224144953e-1,
+        -3.337004338674007801307e-2,
+        1.560544881127388842819e-3,
+        -2.523137093603234562648e-5,
+    ]
+    return _eval_expint_k(A, B, x)
+
+
+def _expint7(x):
+    # x > 64
+    A = [
+        -7.657847078286127362028e-1,
+        6.886192415566705051750e-1,
+        -2.132598113545206124553e-1,
+        3.346107552384193813594e-2,
+        -3.076541477344756050249e-3,
+        1.747119316454907477380e-4,
+        -6.103711682274170530369e-6,
+        1.218032765428652199087e-7,
+        -1.086076102793290233007e-9,
+    ]
+    B = [
+        1.0,
+        -1.888802868662308731041e0,
+        1.066691687211408896850e0,
+        -2.751915982306380647738e-1,
+        3.930852688233823569726e-2,
+        -3.414684558602365085394e-3,
+        1.866844370703555398195e-4,
+        -6.345146083130515357861e-6,
+        1.239754287483206878024e-7,
+        -1.086076102793126632978e-9,
+    ]
+    return _eval_expint_k(A, B, x)
+
+
+def _expi_pos(x):
+    # x > 0
+    _c = _constant_like
+    conds = [(_c(x, 0) < x) & (x <= _c(x, 2))] + [
+        (_c(x, 2 ** i) < x) & (x <= _c(x, 2 ** (i + 1))) for i in range(1, 6)
+    ]
+    return jnp.piecewise(
+        x,
+        conds,
+        [_expint1, _expint2, _expint3, _expint4, _expint5, _expint6, _expint7],
+    )
+
+
+@_wraps(osp_special.expi)
+@api.custom_jvp
+@jit
+def expi(x):
+    (x,) = _promote_args_inexact("expi", x)
+    ret = jnp.piecewise(x, [x < 0], [lambda x: -exp1(-x), _expi_pos])
+    return ret
+
+
+@expi.defjvp
+@jit
+def expi_jvp(primals, tangents):
+    (x,) = primals
+    (x_dot,) = tangents
+    return expi(x), jnp.exp(x) / x * x_dot
+
+
+def _expn1(n, x):
+    # exponential integral En
+    _c = _constant_like
+    x = jnp.array(x)
+    MACHEP = jnp.finfo(x.dtype).eps
+
+    def gamma(x):
+        from jax.scipy.special import gammaln
```

Comment:
I think this import can be removed, because gammaln is defined above in this file